dive_core/perf_metrics_data.cpp

```diff
@@ -223,18 +235,269 @@ PerfMetricsData::PerfMetricsData(std::vector<std::string>          metric_names,
 {
 }
 
-std::unique_ptr<PerfMetricsDataProvider> PerfMetricsDataProvider::Create(
-std::unique_ptr<PerfMetricsData> data)
+class PerfMetricsDataProvider::Correlator
 {
-    if (!data)
-        return nullptr;
+public:
+    using SubmitRef = uint64_t;
+    using CommandBufferRef = uint64_t;
+    using DrawRef = uint64_t;
+    using PatternIndex = size_t;
+    static constexpr PatternIndex kInvalidPatternIndex = std::numeric_limits<PatternIndex>::max();
+
+    void Reset();
+
+    void AnalyzeCommands(const CommandHierarchy&);
+
+    void AnalyzeRecords(const std::vector<PerfMetricsRecord>&);
+
+    std::optional<PatternIndex> MatchOf(size_t index) const;
+
+    size_t GetPatternSize() const { return m_pattern_size; }
+
+    std::optional<DrawRef> GetDrawRef(PatternIndex offset) const
+    {
+        if (offset >= m_draws.size())
+        {
+            return std::nullopt;
+        }
+        return m_draws[offset];
+    }
+    std::optional<PatternIndex> GetPatternIndex(DrawRef ref) const
+    {
+        auto iter = m_draw_to_index.find(ref);
+        if (iter == m_draw_to_index.end())
+        {
+            return std::nullopt;
+        }
+        return iter->second;
+    }
+
+private:
+    static void ExtractDraws(const CommandHierarchy&                 command_hierarchy,
+                             std::vector<uint64_t>&                  out_draws,
+                             std::unordered_map<uint64_t, uint64_t>& out_mapping);
+
+    static bool IsSimilar(size_t n, const PerfMetricsRecord* a, const PerfMetricsRecord* b);
+
+    std::vector<DrawRef>                      m_draws;
+    std::unordered_map<DrawRef, PatternIndex> m_draw_to_index;
+
+    size_t                    m_pattern_size;
+    std::vector<PatternIndex> m_record_to_index;
+};
+
+std::optional<PerfMetricsDataProvider::Correlator::PatternIndex>
+PerfMetricsDataProvider::Correlator::MatchOf(size_t index) const
+{
+
+    if (index >= m_record_to_index.size())
+    {
+        return std::nullopt;
+    }
+    PatternIndex pattern_index = m_record_to_index[index];
+    if (pattern_index >= m_pattern_size)
+    {
+        return std::nullopt;
+    }
+    return pattern_index;
+}
+
+void PerfMetricsDataProvider::Correlator::ExtractDraws(
+const CommandHierarchy&                 command_hierarchy,
+std::vector<uint64_t>&                  out_draws,
+std::unordered_map<uint64_t, uint64_t>& out_mapping)
+{
+    std::unordered_map<uint64_t, uint64_t> draw_to_index;
+
+    std::vector<uint64_t> actual_draws;
```

Comment:
if i understand correctly, the actual_draws contains all drawcalls in the binning pass, and alias_draws contains all drawcalls in the tiled passes? and this func builds a map that is used to map the drawcalls in the PM4 treeview drawcalls to the perf counter?

Could you add some comments to explain this? it might be a bit tricky to understand the logic.

Reply:
That's basically the idea. If the commands are known, it also being used for validating the number of draw call in a frame. There's more information in the command streams we can use to matching draws better, but current approach should work for a large portion of traces.